router/internal/retrytransport/retry_transport.go

```diff
@@ -0,0 +1,118 @@
+package retrytransport
+
+import (
+	"errors"
+	"github.com/cloudflare/backoff"
+	"go.uber.org/zap"
+	"net/http"
+	"strings"
+	"syscall"
+	"time"
+)
+
+var defaultRetryableErrors = []error{
+	syscall.ECONNREFUSED, // "connection refused"
+	syscall.ECONNRESET,   // "connection reset by peer"
+	syscall.ETIMEDOUT,    // "operation timed out"
+	errors.New("i/o timeout"),
+	errors.New("no such host"),
+	errors.New("handshake failure"),
+	errors.New("handshake timeout"),
+	errors.New("timeout awaiting response headers"),
+	errors.New("unexpected EOF"),
+	errors.New("unexpected EOF reading trailer"),
+}
+
+var defaultRetryableStatusCodes = []int{
+	http.StatusBadGateway,
+	http.StatusServiceUnavailable,
+	http.StatusGatewayTimeout,
+	http.StatusTooManyRequests,
+}
+
+type ShouldRetryFunc func(err error, req *http.Request, resp *http.Response) bool
+
+type RetryOptions struct {
+	Enabled       bool
+	MaxRetryCount int
+	Interval      time.Duration
+	MaxDuration   time.Duration
+	OnRetry       func(count int, req *http.Request, resp *http.Response, err error)
+	ShouldRetry   ShouldRetryFunc
+}
+
+type RetryHTTPTransport struct {
+	RoundTripper http.RoundTripper
+	RetryOptions RetryOptions
+	Logger       *zap.Logger
+}
+
+func NewRetryHTTPTransport(roundTripper http.RoundTripper, retryOptions RetryOptions, logger *zap.Logger) *RetryHTTPTransport {
+	return &RetryHTTPTransport{
+		RoundTripper: roundTripper,
+		RetryOptions: retryOptions,
+		Logger:       logger,
+	}
+}
+
+func (rt *RetryHTTPTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+
+	resp, err := rt.RoundTripper.RoundTrip(req)
```

Comment:
I'd add pseudocode after this:
If err is nil and resp status code is 200, return immediately.
No need to create a backoff or do further evaluations if we should retry.